### DISCUSSION TITLE
Requirements for the `url` parameter for `create_engine` if `pool` is specified?

### DISCUSSION THREAD

**Discussion by jfly** [sqlalchemy/sqlalchemy#9648] [Usage Questions] [answered]
We're using `create_engine`'s `pool` parameter to connect to a mysqldb using pymysql. Our code looks something like this:

```python
def get_pymysql_connection():
    return pymysql.connect(
        user='username',
        passwd=get_password(),
        host='db',
        port=3306,
        db=db,
        charset='utf8mb4',
    )

sqlalchemy.create_engine(
    'mysql+pymysql://username:password@db/schema?charset=utf8mb4',
    pool=QueuePool(get_pymysql_connection),
)
```

(The interesting thing going on here is that `get_pymysql_connection` potentially connects with a different password each time it's called.)

Is there any value in specifying a username/password/db in the url we pass to create_engine given that the `QueuePool` has all the smarts for actually making connections to our database? After some playing around and skimming the create_engine source, I think the answer is "no, you really just need a url that'll make [`_parse_url`](https://github.com/sqlalchemy/sqlalchemy/blob/b066c45dc491716a4e34bdeb9a9726bcbc136fb5/lib/sqlalchemy/engine/url.py#L843) happy and will return a valid dialect. I've found that simply specifying `"mysql+pymysql://"` alone is enough to make everything work for us. Is that an acceptable way of using sqlalchemy? Or are there downsides we're not aware of?

Reading the relevant docs, here's what I found:

- [`create_engine`](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine)
  > The [Pool](https://docs.sqlalchemy.org/en/20/core/pooling.html#sqlalchemy.pool.Pool) in turn will establish the first actual DBAPI connection when this request is received. The [create_engine()](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine) call itself does not establish any actual DBAPI connections directly.
- [Database URLs](https://docs.sqlalchemy.org/en/20/core/engines.html#database-urls). Emphasis mine:
  > URLs *typically* include username, password, hostname, database name fields
- [The `pool` parameter](https://docs.sqlalchemy.org/en/20/core/engines.html#sqlalchemy.create_engine.params.pool)
  > If non-None, this pool will be used directly as the underlying connection pool for the engine, bypassing whatever connection parameters are present in the URL argument.

We also found [this documentation on generating dynamic authentication tokens](https://docs.sqlalchemy.org/en/20/core/engines.html#generating-dynamic-authentication-tokens) which looks like it might be another way of achieving what we're doing, but it frankly feels weirder than what we've already done.

**Comment by CaselIT** [comment, marked as answer]
Hi,

No, I don't think there are any downsides to only passing the dialect and dbapi in the url if you pass in the pool.

**Comment by zzzeek** [comment]
the [do_connect](https://docs.sqlalchemy.org/en/20/core/engines.html#fully-replacing-the-dbapi-connect-function) event hook is more modern but has the same effect as passing "creator".   When you pass "creator", the "do_connect" event is skipped.

**Comment by jfly** [comment]
Thanks for the quick responses! It's good to know about that `do_connect` hook, it sounds like it does exactly what we want.
